### ISSUE TITLE
Slow EROFS creation speed with compression

### ISSUE THREAD

**Issue by wszqkzqk** [erofs/erofs-openbenchmark#1]
Hello, I am a user of erofs and I appreciate your work on this project. I have been testing erofs on my Arch Linux system and I found that the erofs creation speed with compression is extremely slow compared to squashfs, which is another popular read-only file system. I would like to report this issue and request some improvement.

I used the following environment and tools to conduct the test:

- CPU: AMD 5800H 8c16t 3200 MHz
- RAM: 16 GB DDR4 3200 MHz
- SSD: TiPlus 5000 PCIe 3.0*4 2 TB
- OS: Arch Linux
- Kernel: 6.4.6
- erofs-utils: 1.6
- squashfs-tools: 4.6.1

I used a folder containing an Arch Linux container with Xfce desktop environment as the source for creating the file systems. The original size of the folder was about 3797448 KiB. I used 1 MiB as the block size for both squashfs and erofs.

I tested the creation speed and compression ratio of squashfs and erofs with different compression algorithms and levels. I also enabled deduplication for squashfs, which is supported by default, but not for erofs, which is still experimental. The results are shown in the following table:

| File system | Compression algorithm | Compression level | Size (KiB) | Compression ratio | Time (s) | Speed (KiB/s) |
|-------------|-----------------------|-------------------|------------|-------------------|----------|---------------|
| squashfs    | zstd                  | default (15)      | 1436510.05 | 37.37%            | 43.99    | 86328.69      |
| squashfs    | zstd                  | 6                 | 1438565.88 | 37.88%            | 16.77    | 226357.23     |
| squashfs    | zstd                  | 3                 | 1493950.67 | 39.58%            | 8.15     | 465885        |
| squashfs    | zstd                  | 1                 | 1602424.36 | 42.20%            | 6.96     | 545443        |
| squashfs    | lz4hc                 | default (9)       | 1951430.30 | 51.39%            | 7.37     | 514986        |
| erofs       | lz4                   | default (1)       | 3519608.00 | **92.68%**        | 10.24    | 370781    |
| erofs       | lz4hc                 | 9                 | 1855644.00 | 48.47%            | **388.97**   | **9762**      |

As you can see, erofs has a much slower creation speed which is **more than 50 times slower** than squashfs with the same compression algorithm and level (lz4hc @ level 9). This might be a significant drawback for using erofs, especially for large data sets.

I wonder if there is any way to improve the erofs creation speed with compression, or if there are any plans to do so in the future. I know that erofs is mainly designed for mobile phones or embedded devices which are of limited performance, but the creation may still be done on a higher performance device, and perhaps parallelization compression or other techniques are still necessary. I think erofs has a lot of potential as a read-only file system, but the slow creation speed may be an obstacle for its adoption.

I also want to acknowledge the advantages of erofs, such as its random access performance and layout design, which are very impressive and innovative. I hope that you can kindly consider my feedback and suggestions, and please let me know if you need any more information or assistance from me.

Thank you for your attention and response.

**Comment by hsiangkao** [comment]
1) erofs-utils is still single-threaded for now due to limited developers and it's not the top priority stuffs (at least for Android scenarios), actually there is a project this summer you could refer to:
https://summer-ospp.ac.cn/org/prodetail/231920038
but I've seen he's somewhat slow on this.

2) if you're a developer, I sincerely invite you (could) spend some time on this if possible.

3) BTW, I don't think this part is tested on the 1MiB pcluster size,
erofs | lz4 | default (1) | 3519608.00 | 92.68% | 10.24 | 370781
-- | -- | -- | -- | -- | -- | --
it seems tested with "-zlz4 -C 4096" instead. Please help correct me if i'm wrong.

**Comment by wszqkzqk** [comment]
> 

I reviewed my terminal's log and found my command:

```bash
time sudo mkfs.erofs -C 1048576 /var/tmp/arch-system.erofs.img /var/tmp/arch-system
```

So I think it's of 1 MiB pcluster size.

**Comment by hsiangkao** [comment]
> > 
> 
> I reviewed my terminal's log and found my command:
> 
> ```shell
> time sudo mkfs.erofs -C 1048576 /var/tmp/arch-system.erofs.img /var/tmp/arch-system
> ```
> 
> So I think it's of 1 MiB pcluster size.

Okay, that is strange since it seems lz4 is nearly incompressable.

I wonder if you could try `-Efragments` at least to match squashfs fragments (squashfs turns it on by default), or `-Efragments,dedupe` for global compressed data deduplication for even smaller images.

**Comment by wszqkzqk** [comment]
> > > 
> > 
> > 
> > I reviewed my terminal's log and found my command:
> > ```shell
> > time sudo mkfs.erofs -C 1048576 /var/tmp/arch-system.erofs.img /var/tmp/arch-system
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So I think it's of 1 MiB pcluster size.
> 
> Okay, that is strange since it seems lz4 is nearly incompressable.
> 
> I wonder if you could try `-Efragments` at least to match squashfs fragments (squashfs turns it on by default), or `-Efragments,dedupe` for global compressed data deduplication for even smaller images.

I tried `-E dedupe` but it was of very low speed (less than 1 MB/s) that I couldn't stand, so I interrupted this test.

**Comment by hsiangkao** [comment]
> > > > 
> > > 
> > > 
> > > I reviewed my terminal's log and found my command:
> > > ```shell
> > > time sudo mkfs.erofs -C 1048576 /var/tmp/arch-system.erofs.img /var/tmp/arch-system
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > So I think it's of 1 MiB pcluster size.
> > 
> > 
> > Okay, that is strange since it seems lz4 is nearly incompressable.
> > I wonder if you could try `-Efragments` at least to match squashfs fragments (squashfs turns it on by default), or `-Efragments,dedupe` for global compressed data deduplication for even smaller images.
> 
> I tried `-E dedupe` but it was of very low speed (less than 1 MB/s) that I couldn't stand, so I interrupted this test.

I'm not quite sure why your CPU is such slow (usually lz4hc,9 can compress 30MB/s for example [1]), maybe you're working on a virtual machine?
Again, it's not a technical issue but a scheduling issue, if you have extra time, you could help us on SIMD improvement of sliding hashing and/or multi-threaded compression.  I have ideas to do but my TODO list is huge.

[1] https://github.com/inikep/lzbench

**Comment by wszqkzqk** [comment]
> > > > > 
> > > > 
> > > > 
> > > > I reviewed my terminal's log and found my command:
> > > > ```shell
> > > > time sudo mkfs.erofs -C 1048576 /var/tmp/arch-system.erofs.img /var/tmp/arch-system
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > So I think it's of 1 MiB pcluster size.
> > > 
> > > 
> > > Okay, that is strange since it seems lz4 is nearly incompressable.
> > > I wonder if you could try `-Efragments` at least to match squashfs fragments (squashfs turns it on by default), or `-Efragments,dedupe` for global compressed data deduplication for even smaller images.
> > 
> > 
> > I tried `-E dedupe` but it was of very low speed (less than 1 MB/s) that I couldn't stand, so I interrupted this test.
> 
> I'm not quite sure why your CPU is such slow (usually lz4hc,9 can compress 30MB/s for example [1]), maybe you're working on a virtual machine? Again, it's not a technical issue but a scheduling issue, if you have extra time, you could help us on SIMD improvement of sliding hashing and/or multi-threaded compression. I have ideas to do but my TODO list is huge.
> 
> [1] https://github.com/inikep/lzbench

> Again, it's not a technical issue but a scheduling issue

Yes, I really understand. I'm just curious why it's even slower than the expected single-threaded result (of a speed lower than `1/16` of squashfs-tools) and whether there are reasons other than lack of parallelization.

> I'm not quite sure why your CPU is such slow (usually lz4hc,9 can compress 30MB/s for example

Yes, the result of squashfs did show this (`514986 / 16 = 32186 KiB/s per thread`), but mkfs.erofs is slower than what we expected to be.

> maybe you're working on a virtual machine?

No, and both squashfs and erofs were tested in the same environment.

> if you have extra time, you could help us on SIMD improvement of sliding hashing and/or multi-threaded compression.

Thank you for your invitation, but I’m afraid I don’t have enough time to directly participate in erofs development. I’m very interested in erofs design and performance, and I hope to contribute to erofs in some way. I think I can help promote erofs in other open source communities, and find more developers and users who might be interested in erofs.

**Comment by hsiangkao** [comment]
`Yes, the result of squashfs did show this (514986 / 16 = 32186 KiB/s per thread), but mkfs.erofs is slower than what we expected to be.`

Could you share your testdata to me?

`but I’m afraid I don’t have enough time to directly participate in erofs development. I’m very interested in erofs design and performance, and I hope to contribute to erofs in some way.`

Honestly, currently EROFS isn't quite lack of big users, but real developers.  I'm more focusing on linux kernel improvements since it's hard to offically backport new killer features to older kernels so it's always top priority for me to improve kernel runtime implementation other than userspace if possible.

**Comment by wszqkzqk** [comment]
> Could you share your testdata to me?

Yes, but I don't know whether you can access this: [arch-system.squashfs, PKU Netdisk](https://disk.pku.edu.cn:443/link/9ACE0EF8342B25336C67DC49D381D904). It's just an Arch Linux container with Xfce desktop, because I think Linux distribution image packaging is an important scenario for read-only file system applications. If you cannot download from this url, you can just create an Arch Linux container with these softwares:

```
amd-ucode
arch-install-scripts
archlinuxcn-keyring
base
base-devel
dosfstools
exfatprogs
exo
f2fs-tools
fcitx5
fcitx5-chinese-addons
fcitx5-configtool
fcitx5-material-color
fcitx5-pinyin-zhwiki
firefox
firefox-i18n-zh-cn
fish
garcon
intel-ucode
lightdm
lightdm-gtk-greeter
lightdm-gtk-greeter-settings
linux
linux-firmware
lvm2
man-db
man-pages
mousepad
nano
neofetch
network-manager-applet
networkmanager
nilfs-utils
nm-connection-editor
noto-fonts-cjk
noto-fonts-emoji
ntfs-3g
parole
pkgfile
ristretto
s-tui
sof-firmware
texinfo
thunar
thunar-archive-plugin
thunar-media-tags-plugin
thunar-volman
ttf-hack
tumbler
xdg-desktop-portal
xdg-user-dirs
xfburn
xfce4-appfinder
xfce4-artwork
xfce4-battery-plugin
xfce4-clipman-plugin
xfce4-cpufreq-plugin
xfce4-cpugraph-plugin
xfce4-dict
xfce4-diskperf-plugin
xfce4-eyes-plugin
xfce4-fsguard-plugin
xfce4-genmon-plugin
xfce4-mailwatch-plugin
xfce4-mount-plugin
xfce4-mpc-plugin
xfce4-netload-plugin
xfce4-notes-plugin
xfce4-notifyd
xfce4-panel
xfce4-power-manager
xfce4-pulseaudio-plugin
xfce4-screensaver
xfce4-screenshooter
xfce4-sensors-plugin
xfce4-session
xfce4-settings
xfce4-smartbookmark-plugin
xfce4-systemload-plugin
xfce4-taskmanager
xfce4-terminal
xfce4-time-out-plugin
xfce4-timer-plugin
xfce4-verve-plugin
xfce4-wavelan-plugin
xfce4-weather-plugin
xfce4-whiskermenu-plugin
xfce4-xkb-plugin
xfconf
xfdesktop
xfsprogs
xfwm4
xfwm4-themes
yay
```

Just create a `pkg.list` and run `sudo pacstrap -c -K /path/to/xxx --needed - < pkg.list` in Arch Linux can generate the test files very similar to mine. Packaging other Linux distribution may also perform similar, I think.🤔

> it's always top priority for me to improve kernel runtime implementation other than userspace if possible

OK, I understand. Thanks for your devotion.

**Comment by hsiangkao** [comment]
> Yes, but I don't know whether you can access this: [arch-system.squashfs, PKU Netdisk](https://disk.pku.edu.cn:443/link/9ACE0EF8342B25336C67DC49D381D904).

The download speed is quite slow on my side: "31.67KB/s", do you have some other way to download this via "wget" directly?
Anyway, the testdata itself is not important, I just want to ensure the absolute costing time on my own machine with the same dataset and do some perf profiling.

**Comment by wszqkzqk** [comment]
> 

Does this work? [arch-system.squashfs, OneDrive](https://chinapku-my.sharepoint.com/:u:/g/personal/2100011770_stu_pku_edu_cn/ESmvIUqfN59Bhkx4PuexZrYBB558R4G9Qm63WaBOuoVSwA?e=aOWyMW). I don't have a public network server for high-speed access, so there may be no wget support.

**Comment by hsiangkao** [comment]
> > 
> 
> Does this work? [arch-system.squashfs, OneDrive](https://chinapku-my.sharepoint.com/:u:/g/personal/2100011770_stu_pku_edu_cn/ESmvIUqfN59Bhkx4PuexZrYBB558R4G9Qm63WaBOuoVSwA?e=aOWyMW). I don't have a public network server for high-speed access, so there may be no wget support.

I fetched later at that time.  Will seek time on playing with it

**Comment by wszqkzqk** [comment]
I did a more detailed test today, including testing different block sizes. I apologize for the inaccurate data I reported to you in my previous comment. I made a mistake in my testing process, which caused me to confuse the lz4 and lz4hc algorithms in squashfs, resulting in a shorter compression time than expected for lz4hc. I corrected this in today’s test. In addition, erofs compressed with lz4 algorithm showed an abnormally high compression ratio in my previous test, which did not appear in today’s test. I did not keep the original files from my previous test, so I have no way to find out the reason. Today’s test data is still an Arch Linux container with Xfce, very similar to my previous test, except that I kept the boot partition. All the test data and logs from today are archived. Here are the test results:


|File system | Block Size (KiB) | Compression algorithm | Compression level | Size (KiB) | Compression ratio | Time (s) | Speed (KiB/s)|
| :------:    | :------: | :------: | :------:    | :------: | :----:  | :----:     | :----------:  |
| squashfs    | 4        | zstd  | 1              | 1911360 | 48.89%     | 7.42      | 526888        |
| squashfs    | 128      | zstd  | 1              | 1740716 | 44.53%     | 6.11      | 639854        |
| squashfs    | 1024     | zstd  | 1              | 1715208 | 43.87%     | 6.04      | 647270        |
| squashfs    | 4        | zstd  | 3              | 1886500 | 48.25%     | 7.25      | 539243        |
| squashfs    | 128      | zstd  | 3              | 1682224 | 43.03%     | 6.21      | 629551        |
| squashfs    | 1024     | zstd  | 3              | 1606452 | 41.09%     | 7.26      | 538500        |
| squashfs    | 4        | zstd  | 6              | 1853852 | 47.42%     | 7.65      | 511047        |
| squashfs    | 128      | zstd  | 6              | 1619220 | 41.42%     | 7.41      | 527599        |
| squashfs    | 1024     | zstd  | 6              | 1551152 | 39.67%     | 15.36     | 254526        |
| squashfs    | 4        | zstd  | default (15)   | 1800144 | 46.05%     | 38.99     | 100269        |
| squashfs    | 128      | zstd  | default (15)   | 1540216 | 39.40%     | 46.63     | 83841         |
| squashfs    | 1024     | zstd  | default (15)   | 1531772 | 39.18%     | 43.62     | 89627         |
| squashfs    | 4        | lz4hc | default (9)    | 2196376 | 56.18%     | 12.16     | 321505        |
| squashfs    | 128      | lz4hc | default (9)    | 1830572 | 46.82%     | 30.59     | 127803        |
| squashfs    | 1024     | lz4hc | default (9)    | 1789976 | 45.79%     | 41.73     | 93685         |
| squashfs    | 4        | lz4   | default (9)    | 2310164 | 59.09%     | 7.69      | 508389        |
| squashfs    | 128      | lz4   | default (9)    | 2078612 | 53.17%     | 6.07      | 644071        |
| squashfs    | 1024     | lz4   | default (9)    | 2063612 | 52.78%     | 6.39      | 611817        |
| erofs       | 4        | lz4hc | 9              | 2225036 | 56.91%     | 76.71     | 50964         |
| erofs       | 128      | lz4hc | 9              | 1995220 | 51.04%     | 261.54    | 14948         |
| erofs       | 1024     | lz4hc | 9              | 1983248 | 50.72%     | 865.25    | 4518          |
| erofs       | 4        | lz4   | default (1)    | 2374256 | 60.73%     | 16.33     | 239407        |
| erofs       | 128      | lz4   | default (1)    | 2239728 | 57.29%     | 16.49     | 237084        |
| erofs       | 1024     | lz4   | default (1)    | 2236176 | 57.20%     | 17.88     | 218653        |

In addition, I find that:

* The compression ratio of zstd algorithm with low compression level is sensitive to block size, and the benefit of increasing block size from 128 KiB to 1 MiB is still significant; in other cases, only increasing block size from 4 KiB to 128 KiB has obvious benefit
* The compression ratio of **lz4hc** algorithm is insensitive to block size, but the compression speed is very sensitive to block size, and increasing **block size** will seriously affect the compression **speed**
* This test still shows that EROFS has a slow creation speed, and it is more than 20 times slower than squashfs when using lz4hc algorithm with compression level 9 and 1 MiB block size
* Due to IO bottleneck, this test cannot reflect much for squashfs’s compression speed when using zstd algorithm with low compression level or lz4 algorithm

[squashfs-erofs-comparision.log](https://github.com/erofs/erofs-openbenchmark/files/12207375/squashfs-erofs-comparision-detail.log)

**Comment by hsiangkao** [comment]
Did you use `-Efragments` for your test results?  Even fragment support is experimental for now, but even after it drops from experimental, EROFS won't enable it by default anyway since it impacts performance and performance is always top priority for EROFS targetted scenarios.

But if you don't enable `-Efragments`, it's unfair comparsion on the image size on my side.  As for your compression speed, that is totally another story as I explained before.

**Comment by wszqkzqk** [comment]
> Did you use `-Efragments` for your test results? Even fragment support is experimental for now, but even after it drops from experimental, EROFS won't enable it by default anyway since it impacts performance and performance is always top priority for EROFS targetted scenarios.
> 
> But if you don't enable `-Efragments`, it's unfair comparsion on the image size on my side. As for your compression speed, that is totally another story as I explained before.

But will that let the compression progress even slower?🤔

**Comment by hsiangkao** [comment]
> > Did you use `-Efragments` for your test results? Even fragment support is experimental for now, but even after it drops from experimental, EROFS won't enable it by default anyway since it impacts performance and performance is always top priority for EROFS targetted scenarios.
> > But if you don't enable `-Efragments`, it's unfair comparsion on the image size on my side. As for your compression speed, that is totally another story as I explained before.
> 
> But will that let the compression progress even slower?🤔

Again, that is another story. Creation speed can be improved if I or other developers have more time but here you need to make a fair comparsion first?

**Comment by hsiangkao** [comment]
> > > Did you use `-Efragments` for your test results? Even fragment support is experimental for now, but even after it drops from experimental, EROFS won't enable it by default anyway since it impacts performance and performance is always top priority for EROFS targetted scenarios.
> > > But if you don't enable `-Efragments`, it's unfair comparsion on the image size on my side. As for your compression speed, that is totally another story as I explained before.
> > 
> > 
> > But will that let the compression progress even slower?🤔
> 
> Again, that is another story. Creation speed can be improved if I or other developers have more time but here you need to make a fair comparsion first?

Or if you want to just report a creation speed issue.
Ok, well received. As I told you before, since current Android vendors don't care much about this compared with runtime performance (that is why they don't put more manpower on the issue you reported as well, they don't care about 1 MiB compression unit too), but I think I will improve it when I have more time to look into that?

**Comment by wszqkzqk** [comment]
OK. I added more tests with `fragments` enabled for erofs and `no-duplicates` enabled for squashfs for more fair comparison. It seems that erofs's compression ratio is better.

|File system | Block Size (KiB) | Compression algorithm | Compression level | Size (KiB) | Compression ratio | Time (s) | Speed (KiB/s)|
| :------:    | :------: | :------: | :------:    | :------: | :----:  | :----:     | :----------:  |
| squashfs (no-duplicates) | 4        | lz4hc | default (9)    | 2239452 | 57.28%     | 12.09      | 323367        |
| squashfs (no-duplicates) | 128      | lz4hc | default (9)    | 1865724 | 47.72%     | 29.79     | 131236        |
| squashfs (no-duplicates) | 1024     | lz4hc | default (9)    | 1823560 | 46.64%     | 40.18     | 97300         |
| squashfs (no-duplicates) | 4        | lz4   | default (9)    | 2354264 | 60.22%     | 8.88      | 440260        |
| squashfs (no-duplicates) | 128      | lz4   | default (9)    | 2115096 | 54.10%     | 5.95      | 657061        |
| squashfs (no-duplicates) | 1024     | lz4   | default (9)    | 2098692 | 53.68%     | 6.25      | 625522        |
| erofs (fragments) | 4        | lz4hc | 9              | 2132044 | 54.53%     | 77.94    | 50161         |
| erofs (fragments) | 128      | lz4hc | 9              | 1842848 | 47.14%     | 330.58   | 11826         |
| erofs (fragments) | 1024     | lz4hc | 9              | 1815460 | 46.44%     | 31.68    | 123406        |
| erofs (fragments) | 4        | lz4   | default (1)    | 2285184 | 58.45%     | 16.04    | 243735        |
| erofs (fragments) | 128      | lz4   | default (1)    | 2102400 | 53.78%     | 21.02    | 185990        |
| erofs (fragments) | 1024     | lz4   | default (1)    | 2084988 | 53.33%     | 27.49    | 142216        |

**Comment by hsiangkao** [comment]
> OK. I added more tests with `fragments` enabled for erofs and `no-duplicates` enabled for squashfs for more fair comparison. It seems that erofs's compression ratio is better.
> 
> File system	Block Size (KiB)	Compression algorithm	Compression level	Size (KiB)	Compression ratio	Time (s)	Speed (KiB/s)
> squashfs (no-duplicates)	4	lz4hc	default (9)	2239452	57.28%	12.09	323367
> squashfs (no-duplicates)	128	lz4hc	default (9)	1865724	47.72%	29.79	131236
> squashfs (no-duplicates)	1024	lz4hc	default (9)	1823560	46.64%	40.18	97300
> squashfs (no-duplicates)	4	lz4	default (9)	2354264	60.22%	8.88	440260
> squashfs (no-duplicates)	128	lz4	default (9)	2115096	54.10%	5.95	657061
> squashfs (no-duplicates)	1024	lz4	default (9)	2098692	53.68%	6.25	625522
> erofs (fragments)	4	lz4hc	9	2132044	54.53%	77.94	50161
> erofs (fragments)	128	lz4hc	9	1842848	47.14%	330.58	11826
> erofs (fragments)	1024	lz4hc	9	1815460	46.44%	31.68	123406
> erofs (fragments)	4	lz4	default (1)	2285184	58.45%	16.04	243735
> erofs (fragments)	128	lz4	default (1)	2102400	53.78%	21.02	185990
> erofs (fragments)	1024	lz4	default (1)	2084988	53.33%	27.49	142216

squashfs only supports file-level deduplication but it's nearly 20-year project.  I do think as the time goes by, EROFS will support more wider scenarios.  Anyway, your request is well received, and the current issue is discussed.

**Comment by hsiangkao** [comment]
currently the upstream erofs-utils already supports with multi-threaded compression, but it still doesn't support with some option anyway.
This feature will be shipped in erofs-utils v1.8 soon.

**Comment by hsiangkao** [comment]
erofs-utils 1.8 shipped, `-Eall-framgents`, `-Eztailpacking` and `(none)` now support multi-threaded compression.
As for `-Efragments` and `-Ededupe` feature, it might support in the future versions (if my time permits or other developer helps).
I'm closing this issue now.
